File: 0567-permutation-in-string/0567-permutation-in-string.java
class Solution {
    public boolean compare(int [] pc,int [] sc){
        for(int i = 0;i < 26;i++){
            if(pc[i] != sc[i])
                return false;
        }
        return true;
    }
    
    public boolean checkInclusion(String s1, String s2) {
        String p,s;
          p = s1;
          s = s2;
        
        
        int pl = p.length(), sl = s.length();
        
        if(pl > sl)return false;
        
        int pc[] = new int[26];
        int sc[] = new int[26];
        
        
        for(int i =0;i < pl;i++){
            char c = p.charAt(i);
            pc[c-'a']++;
            
            char d = s.charAt(i);
            sc[d-'a']++;
        }
        
        if(compare(pc,sc)){
            return true;
        }
        
        int left = 0, right = pl;
        while(right < sl){
            char lc = s.charAt(left);
            char rc = s.charAt(right);
            
            sc[lc-'a']--;
            sc[rc-'a']++;
            
            if(compare(pc,sc)){
                return true;   
            }
            left++;right++;
        }
        return false;
    }
}

File: 0567-permutation-in-string/0567-permutation-in-string.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if len(s2) < len(s1):
            return False

        ct1 = [0 for i in range(26)]
        ct2 = [0 for i in range(26)]

        for c in s1:
            id = ord(c) - ord("a")
            ct1[id] += 1

        i2 = 0

        for i in range(len(s1)):
            id = ord(s2[i]) - ord("a")
            ct2[id] += 1

        if ct1 == ct2:
            return True

        for i in range(1, (len(s2) - len(s1) + 1)):
            c = s2[i - 1]
            id = ord(c) - ord("a")

            ct2[id] -= 1

            d = s2[i + len(s1) - 1]
            id1 = ord(d) - ord("a")
            ct2[id1] += 1

            if ct2 == ct1:
                return True

        return False
